nemo/core/nemo_model.py

```diff
@@ -0,0 +1,240 @@
+
+
+class NeMoModel(NeuralModule):
+    """Abstract class representing NeMoModel.
+    A NeMoModel is a kind of neural module which contains other neural modules and logic inside.
+    It typically represents a whole neural network and requires only connections with data layer and loss
+    modules for training.
+
+    The same NeMoModel could be used in training, evaluation or inference regimes. It should adjust itself
+    accordingly.
+    """
+
+    def __call__(self, **kwargs):
+        if self._operation_mode == OperationMode.training or self.operation_mode == OperationMode.both:
+            return self.train_graph(**kwargs)
+
+        else:
+            return self.eval_graph(**kwargs)
+
+    @classmethod
+    @abstractmethod
+    def from_pretrained(cls, model_info, local_rank: int = 0, referesh_cache: bool = False) -> NeuralModule:
+        """
+        Instantiates NeMoModel from pretrained checkpoint. Can do so from file on disk or from the NVIDIA NGC.
+        Args:
+            model_info: Either path to ".nemo" file or a valid NGC Model name
+            local_rank: on which GPU to instantiate.
+            referesh_cache: If set to True, then when fetching from clould, this will re-fetch the file
+                from clould even if it is  already found in a cache locally.
+
+        Returns:
+            NeMoModel instance
+        """
+        if isinstance(model_info, str) and model_info.endswith(".nemo"):
+            nemo_file_folder, to_delete = cls.__unpack_nemo_file(path2file=model_info)
+            configuration_file = path.join(nemo_file_folder, 'module.yaml')
+            instance = cls.import_from_config(config_file=configuration_file)
+            for module in instance.modules:
+                module_checkpoint = path.join(nemo_file_folder, module.__class__.__name__ + ".pt")
+                module.restore_from(path=module_checkpoint, local_rank=local_rank)
+            shutil.rmtree(to_delete)
+            return instance
+        else:
+            raise NotImplemented("Generic from_pretrained from cloud is not implemented")
+
+    def save_to(self, output_file_name: str, output_folder: str = None, optimize_for_deployment: bool = False) -> str:
+        """
+        Saves NeMoModel to .nemo file. This file will contain:
+            * weights of all NeuralModule instances inside the model
+            * Yaml file with configuration
+            * Yaml files with topologies and configuration of training and (if applicable) eval graphs
+        Args:
+            output_file_name: filename, something like nemomodel.nemo
+            output_folder: folder where to save output_file_name. If None (default) current folder will be used.
+            optimize_for_deployment: will optimize for deployment by trying to export modules to .onnx format and
+                skipping training graph.
+
+        Returns:
+            None
+        """
+
+        def make_nemo_file_from_folder(filename, source_dir):
```

Comment:
private method? do we consider user to call that method directly?

Reply:
how about making the names _symmetrical_?

__unpack_nemo_file -> __pack_nemo_file ?

Reply:
good point - made it private